src/main/scala/temple/DSL/Semantics/package.scala

```diff
@@ -0,0 +1,383 @@
+    val metadatas  = mutable.ListBuffer[ServiceMetadata]()
+    val structs    = mutable.LinkedHashMap[String, StructBlock]()
+    entries.foreach {
+      case Entry.Attribute(key, dataType, annotations) =>
+        attributes.safeInsert(key -> parseAttributeType(dataType, annotations)(KeyName(key)))
+      case Entry.Metadata(metaKey, args) => metadatas += parseServiceMetadata(metaKey, args)
+      case DSLRootItem(key, tag, entries) =>
+        tag match {
+          case "struct" => structs.safeInsert(key -> parseStructBlock(entries)(BlockContext(key, tag, context)))
+          case tag      => fail(s"Unknown block type $tag for $key in $context")
+        }
+    }
+    ServiceBlock(attributes.to(ListMap), metadatas.toList, structs.to(ListMap))
+  }
+
+  // TODO: these methods are very similar, can they be merged?
+
+  private def parseProjectBlock(entries: List[Entry])(implicit context: BlockContext): ProjectBlock = entries map {
+    case Entry.Metadata(metaKey, args) => parseProjectMetadata(metaKey, args)
+    case otherEntry                    =>
+      // TODO: should this hardcode ${context.tag} as Project? Can we get a link to the relevant part of the docs
+      fail(s"Found ${otherEntry.typeName} in ${context.tag} block (${context.block}): `$otherEntry`")
+  }
+
+  private def parseTargetBlock(entries: List[Entry])(implicit context: BlockContext): TargetBlock = entries map {
+    case Entry.Metadata(metaKey, args) => parseTargetMetadata(metaKey, args)
+    case otherEntry =>
+      fail(s"Found ${otherEntry.typeName} in ${context.tag} block (${context.block}): `$otherEntry`")
+  }
+
+  private def parseStructBlock(entries: List[Entry])(implicit context: BlockContext): StructBlock = {
+    val attributes = mutable.LinkedHashMap[String, AttributeType]()
+    entries.foreach {
+      case Entry.Attribute(key, dataType, annotations) =>
+        attributes.safeInsert(key -> parseAttributeType(dataType, annotations)(KeyName(key)))
+      case otherEntry =>
+        // TODO: should this hardcode ${context.tag} as Project? Can we get a link to the relevant part of the docs
+        fail(s"Found ${otherEntry.typeName} in ${context.tag} block (${context.block}): `$otherEntry`")
+    }
+    StructBlock(attributes.toMap)
+  }
+
+  /**
+    * Turns an AST of a Templefile into a semantic representation
+    * @param templefile the AST parsed from the Templefile
+    * @return A semantic representation of the project, as well as all the targets and services, defined in the
+    *         Templefile
+    * @throws SemanticParsingException when there is no project information, as well as when any of the definitions are
+    *                                  malformed
+    */
+  def parseSemantics(templefile: Syntax.Templefile): Semantics.Templefile = {
+    var projectNameBlock: Option[(String, ProjectBlock)] = None
+
+    val targets  = mutable.LinkedHashMap[String, TargetBlock]()
+    val services = mutable.LinkedHashMap[String, ServiceBlock]()
```

Comment:
`LinkedHashMap` used to maintain insertion order, I should add a comment to explain that